gateway-sp-comms/src/single_sp.rs

```diff
@@ -2052,3 +2057,110 @@ mod probes {
+            let (recv_tx, recv) = mpsc::unbounded_channel();
+            (Self { log, packets_sent: Vec::new(), recv }, recv_tx)
+        }
+    }
+
+    #[async_trait]
+    impl InnerSocket for ChannelInnerSocket {
+        fn log(&self) -> &Logger {
+            &self.log
+        }
+
+        fn discovery_addr(&self) -> SocketAddrV6 {
+            unimplemented!()
+        }
+
+        async fn send(
+            &mut self,
+            data: &[u8],
+        ) -> Result<(), SingleSpHandleError> {
+            self.packets_sent.push(data.into());
+            Ok(())
+        }
+
+        async fn recv(&mut self) -> SingleSpMessage {
+            self.recv.recv().await.unwrap()
+        }
+    }
+
+    #[tokio::test]
+    async fn rpc_call_one_attempt_times_out_while_receiving_host_request_updates(
+    ) {
+        let (sp_addr_tx, _sp_addr_rx) = watch::channel(None);
+        let (_cmds_tx, cmds_rx) = mpsc::channel(128);
+        let (socket, socket_tx) =
+            ChannelInnerSocket::new(Logger::root(slog::Discard, slog::o!()));
+        let mut inner = Inner::new(
+            socket,
+            sp_addr_tx,
+            1,
+            Duration::from_millis(200),
+            cmds_rx,
+        );
+
+        // Spawn a task that emulates the SP sending host phase 2 requests on a
+        // frequency that's higher than our timeout (we'll do 20ms, so 10x
+        // higher).
+        tokio::spawn(async move {
+            let req = SingleSpMessage::HostPhase2Request(HostPhase2Request {
+                hash: [0; 32],
+                offset: 0,
+                data_sent: 0,
+                received: Instant::now(),
+            });
+            loop {
+                if socket_tx.send(req.clone()).is_err() {
+                    return;
+                }
+                tokio::time::sleep(Duration::from_millis(20)).await;
+            }
+        });
+
+        // Call `rpc_call_one_attempt`; this should time out in ~200ms, but
+        // we'll be generous to overloaded systems like CI and give it 10x that
+        // time.
+        let start = Instant::now();
+        match tokio::time::timeout(
+            Duration::from_secs(2),
+            inner.rpc_call_one_attempt(0, b"dummy"),
+        )
+        .await
+        {
+            // rpc_call_one_attempt timed itself out as expected
+            Ok(Ok(None)) => {
+                assert!(
+                    start.elapsed() >= Duration::from_millis(200),
+                    "rpc_call_one_attempt returned after {:?} \
+                     (we expected a timeout after 200ms",
+                    start.elapsed(),
+                );
+            }
+            Ok(Ok(Some(value))) => panic!("unexpected response {value:?}"),
+            Ok(Err(err)) => panic!("unexpected error {err}"),
+            Err(_elapsed) => {
+                panic!(
+                    "rpc_call_one_attempt failed to time out \
+                     (expected timeout after 200ms, waited 2000ms)"
```

Comment:
Without the corresponding change to timeout handling, this is the test failure I saw, which matches the behavior we saw on rack2 (where our "2 second timeout" here is playing the role of wicketd - seeing timeouts when sending requests to specific SPs).